src/main/java/ch/njol/skript/events/EvtDamage.java

```diff
@@ -38,35 +38,53 @@
  */
 @SuppressWarnings("unchecked")
 public class EvtDamage extends SkriptEvent {
+
 	static {
-		Skript.registerEvent("Damage", EvtDamage.class, EntityDamageEvent.class, "damag(e|ing) [of %entitydata%]")
+		Skript.registerEvent("Damage", EvtDamage.class, EntityDamageEvent.class, "damag(e|ing) [of %entitydata%] [by %entitydata%]")
 				.description("Called when an entity receives damage, e.g. by an attack from another entity, lava, fire, drowning, fall, suffocation, etc.")
-				.examples("on damage:", "on damage of a player:")
-				.since("1.0");
+				.examples("on damage:", "on damage of a player:", "on damage of player by zombie:")
+				.since("1.0, INSERT VERSION for by entitydata");
 	}
 	
 	@Nullable
-	private Literal<EntityData<?>> types;
+	private Literal<EntityData<?>> ofTypes, byTypes;
 	
 	@Override
-	public boolean init(final Literal<?>[] args, final int matchedPattern, final ParseResult parser) {
-		types = (Literal<EntityData<?>>) args[0];
+	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parser) {
+		ofTypes = (Literal<EntityData<?>>) args[0];
+		byTypes = (Literal<EntityData<?>>) args[1];
 		return true;
 	}
 	
 	@Override
-	public boolean check(final Event evt) {
-		final EntityDamageEvent e = (EntityDamageEvent) evt;
-		if (!checkType(e.getEntity()))
+	public boolean check(Event evt) {
+		EntityDamageEvent e = (EntityDamageEvent) evt;
+		if (evt instanceof EntityDamageByEntityEvent) {
+			EntityDamageByEntityEvent event = (EntityDamageByEntityEvent) evt;
+			if (!checkDamager(event.getDamager()))
+				return false;
+		} else if (byTypes != null) return false;
+		if (!checkDamaged(e.getEntity()))
```

Comment:
```suggestion
		if (byTypes != null || !checkDamaged(e.getEntity()))
```
Not to nitpick but just to keep formatting consistent / save space >:)

This should be it ;D

Reply:
shall do :DDD

Reply:
wait wouldn't this break it? as it doesn't return true inside the if (evt instanceof whatever), it just returns false, the byTypes could still be not null

Reply:
Oops - my bad. I've edited my suggestion to be clearer.

Reply:
I see, shall do !